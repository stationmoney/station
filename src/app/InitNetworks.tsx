import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
