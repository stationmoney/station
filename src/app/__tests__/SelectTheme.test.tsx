import React from "react"
import SelectTheme from "../sections/SelectTheme"
import { RecoilRoot } from "recoil"
import { NetworksProvider } from "../InitNetworks"
import { render } from "@testing-library/react"
import { mockNetworks } from "./__mocks__/Networks.mock"

function renderComponent() {
  type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = {} as jest.Mocked<TokenFilter>

  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <RecoilRoot>
        <SelectTheme />
      </RecoilRoot>
    </NetworksProvider>
  )
}

jest.mock("../../data/wallet", () => {
  const mockUseNetworkName = () => {
    return "testnet"
  }

  const mockUseAddress = () => {
    return "terra111111111111111111111111111111111111111"
  }

  const mockUseChainID = () => {
    return "pisco-1"
  }

  const mockUseNetwork = () => {
    return mockNetworks
  }

  const mockUseDelegations = () => {
    return []
  }

  return {
    useNetworkName: mockUseNetworkName,
    useAddress: mockUseAddress,
    useChainID: mockUseChainID,
    useNetwork: mockUseNetwork,
    useDelegations: mockUseDelegations,
  }
})

jest.mock("../../data/settings/Theme", () => {
  const mockUseValidateTheme = () => {}

  return {
    useValidateTheme: mockUseValidateTheme,
  }
})

const mockSetTheme = jest.fn()

jest.mock("../../data/settings/Theme", () => {
  const mockUseThemeState = () => ["dark", mockSetTheme]
  const mockUseValidateTheme = () => {}

  return {
    useThemeState: mockUseThemeState,
    useValidateTheme: mockUseValidateTheme,
  }
})

describe("SelectTheme component matches snapshots", () => {
  it("matches original component", () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
