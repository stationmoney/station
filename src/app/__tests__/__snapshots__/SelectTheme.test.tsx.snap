// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectTheme component matches snapshots matches original component 1`] = `
<DocumentFragment>
  <div
    class="grid"
    style="gap: 20px;"
  >
    <div
      class="grid"
      style="gap: 28px; grid-template-columns: repeat(2, 1fr);"
    >
      <button
        class="component disabled locked"
        type="button"
      >
        <div
          class="grid"
          style="gap: 4px;"
        >
          <div
            class="flex start"
            style="gap: 8px;"
          >
            <div
              class="flex track"
            >
              <span
                class="indicator"
              />
            </div>
            Light
          </div>
          <div
            class="wrapper"
          >
            <div
              class="flex preview"
            >
              <svg>
                preview.svg
              </svg>
            </div>
            <div
              class="column unlock"
              style="gap: 4px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="LockIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
              <small>
                Stake {{amount}} Luna to unlock
              </small>
            </div>
          </div>
        </div>
      </button>
      <button
        class="component disabled locked"
        type="button"
      >
        <div
          class="grid"
          style="gap: 4px;"
        >
          <div
            class="flex start"
            style="gap: 8px;"
          >
            <div
              class="flex track"
            >
              <span
                class="indicator"
              />
            </div>
            Dark
          </div>
          <div
            class="wrapper"
          >
            <div
              class="flex preview"
            >
              <svg>
                preview.svg
              </svg>
            </div>
            <div
              class="column unlock"
              style="gap: 4px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="LockIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
              <small>
                Stake {{amount}} Luna to unlock
              </small>
            </div>
          </div>
        </div>
      </button>
      <button
        class="component disabled locked"
        type="button"
      >
        <div
          class="grid"
          style="gap: 4px;"
        >
          <div
            class="flex start"
            style="gap: 8px;"
          >
            <div
              class="flex track"
            >
              <span
                class="indicator"
              />
            </div>
            Blossom
          </div>
          <div
            class="wrapper"
          >
            <div
              class="flex preview"
            >
              <svg>
                preview.svg
              </svg>
            </div>
            <div
              class="column unlock"
              style="gap: 4px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="LockIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
              <small>
                Stake {{amount}} Luna to unlock
              </small>
            </div>
          </div>
        </div>
      </button>
      <button
        class="component disabled locked"
        type="button"
      >
        <div
          class="grid"
          style="gap: 4px;"
        >
          <div
            class="flex start"
            style="gap: 8px;"
          >
            <div
              class="flex track"
            >
              <span
                class="indicator"
              />
            </div>
            Moon
          </div>
          <div
            class="wrapper"
          >
            <div
              class="flex preview"
            >
              <svg>
                preview.svg
              </svg>
            </div>
            <div
              class="column unlock"
              style="gap: 4px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="LockIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
              <small>
                Stake {{amount}} Luna to unlock
              </small>
            </div>
          </div>
        </div>
      </button>
      <button
        class="component disabled locked"
        type="button"
      >
        <div
          class="grid"
          style="gap: 4px;"
        >
          <div
            class="flex start"
            style="gap: 8px;"
          >
            <div
              class="flex track"
            >
              <span
                class="indicator"
              />
            </div>
            Whale
          </div>
          <div
            class="wrapper"
          >
            <div
              class="flex preview"
            >
              <svg>
                preview.svg
              </svg>
            </div>
            <div
              class="column unlock"
              style="gap: 4px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="LockIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
              <small>
                Stake {{amount}} Luna to unlock
              </small>
            </div>
          </div>
        </div>
      </button>
      <button
        class="component disabled locked"
        type="button"
      >
        <div
          class="grid"
          style="gap: 4px;"
        >
          <div
            class="flex start"
            style="gap: 8px;"
          >
            <div
              class="flex track"
            >
              <span
                class="indicator"
              />
            </div>
            Madness
          </div>
          <div
            class="wrapper"
          >
            <div
              class="flex preview"
            >
              <svg>
                preview.svg
              </svg>
            </div>
            <div
              class="column unlock"
              style="gap: 4px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="LockIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
              <small>
                Stake {{amount}} Luna to unlock
              </small>
            </div>
          </div>
        </div>
      </button>
    </div>
    <div
      class="flex info"
      style="gap: 4px;"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="InfoIcon"
        focusable="false"
        style="font-size: 18px;"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
        />
      </svg>
      Preview is available if wallet is disconnected
    </div>
  </div>
</DocumentFragment>
`;
